library/psa_crypto.c

```diff
@@ -3131,7 +3156,733 @@ psa_status_t psa_asymmetric_decrypt( mbedtls_svc_key_id_t key,
+                            const uint8_t *hash, size_t hash_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot;
+
+    /* Check that start has not been previously called, or operation has not
+     * previously errored. */
+    if( operation->id != 0 || operation->error_occurred )
+        return( PSA_ERROR_BAD_STATE );
+
+    status = psa_sign_verify_check_alg( 0, alg );
+    if( status != PSA_SUCCESS )
+    {
+        operation->error_occurred = 1;
+        return status;
+    }
+
+    status = psa_get_and_lock_key_slot_with_policy( key, &slot,
+                                                    PSA_KEY_USAGE_SIGN_HASH,
+                                                    alg );
+
+    if( status != PSA_SUCCESS )
+        goto exit;
+
+    if( !PSA_KEY_TYPE_IS_KEY_PAIR( slot->attr.type ) )
+    {
+        status = PSA_ERROR_INVALID_ARGUMENT;
+        goto exit;
+    }
+
+    psa_key_attributes_t attributes = {
+        .core = slot->attr
+    };
+
+    /* Ensure ops count gets reset, in case of operation re-use. */
+    operation->num_ops = 0;
+
+    status = psa_driver_wrapper_sign_hash_start( operation, &attributes,
+                                                 slot->key.data,
+                                                 slot->key.bytes, alg,
+                                                 hash, hash_length );
+exit:
+
+    if( status != PSA_SUCCESS )
+    {
+        operation->error_occurred = 1;
+        psa_sign_hash_abort_internal( operation );
+    }
+
+    unlock_status = psa_unlock_key_slot( slot );
+
+    if( unlock_status != PSA_SUCCESS )
+        operation->error_occurred = 1;
+
+    return( ( status == PSA_SUCCESS ) ? unlock_status : status );
+}
+
+
+psa_status_t psa_sign_hash_complete(
+                            psa_sign_hash_interruptible_operation_t *operation,
+                            uint8_t *signature, size_t signature_size,
+                            size_t *signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    *signature_length = 0;
+
+    /* Check that start has been called first, and that operation has not
+     * previously errored. */
+    if( operation->id == 0 || operation->error_occurred )
+    {
+        status = PSA_ERROR_BAD_STATE;
+        goto exit;
+    }
+
+    /* Immediately reject a zero-length signature buffer. This guarantees that
+     * signature must be a valid pointer. */
+    if( signature_size == 0 )
+    {
+        status = PSA_ERROR_BUFFER_TOO_SMALL;
+        goto exit;
+    }
+
+    status = psa_driver_wrapper_sign_hash_complete( operation, signature,
+                                                    signature_size,
+                                                    signature_length );
+exit:
+
+    /* Update ops count with work done. */
+    operation->num_ops = psa_driver_wrapper_sign_hash_get_num_ops( operation );
```

Comment:
ok, agreed. I was trying to make sure that no work would end up being uncaptured even in case of error, but you are quite correct, this cannot happen, we are just calling it when no work was done in error cases.